gpu/containers/src/initialization.cpp

```diff
@@ -1,247 +1,358 @@
-        printf("  Maximum memory pitch:                          %u bytes\n", (int)prop.memPitch);
-        printf("  Texture alignment:                             %u bytes\n", (int)prop.textureAlignment);
-
-        printf("  Concurrent copy and execution:                 %s with %d copy engine(s)\n", (prop.deviceOverlap ? "Yes" : "No"), prop.asyncEngineCount);
-        printf("  Run time limit on kernels:                     %s\n", prop.kernelExecTimeoutEnabled ? "Yes" : "No");
-        printf("  Integrated GPU sharing Host Memory:            %s\n", prop.integrated ? "Yes" : "No");
-        printf("  Support host page-locked memory mapping:       %s\n", prop.canMapHostMemory ? "Yes" : "No");
-
-        printf("  Concurrent kernel execution:                   %s\n", prop.concurrentKernels ? "Yes" : "No");
-        printf("  Alignment requirement for Surfaces:            %s\n", prop.surfaceAlignment ? "Yes" : "No");
-        printf("  Device has ECC support enabled:                %s\n", prop.ECCEnabled ? "Yes" : "No");
-        printf("  Device is using TCC driver mode:               %s\n", prop.tccDriver ? "Yes" : "No");
-        printf("  Device supports Unified Addressing (UVA):      %s\n", prop.unifiedAddressing ? "Yes" : "No");
-        printf("  Device PCI Bus ID / PCI location ID:           %d / %d\n", prop.pciBusID, prop.pciDeviceID );
-        printf("  Compute Mode:\n");
-        printf("      %s \n", computeMode[prop.computeMode]);
-    }
-    
-    printf("\n");    
-    printf("deviceQuery, CUDA Driver = CUDART");
-    printf(", CUDA Driver Version  = %d.%d", driverVersion / 1000, driverVersion % 100);
-    printf(", CUDA Runtime Version = %d.%d", runtimeVersion/1000, runtimeVersion%100);
-    printf(", NumDevs = %d\n\n", count);                
-    fflush(stdout);
+  // Defines for GPU Architecture types (using the SM version to determine the # of
+  // cores per SM
+  struct SMtoCores {
+    int SM; // 0xMm (hexadecimal notation), M = SM Major version, and m = SM minor
+            // version
+    int Cores;
+  };
+
+  SMtoCores gpuArchCoresPerSM[] = {{0x10, 8},
+                                   {0x11, 8},
+                                   {0x12, 8},
+                                   {0x13, 8},
+                                   {0x20, 32},
+                                   {0x21, 48},
+                                   {0x30, 192},
+                                   {0x35, 192},
+                                   {0x50, 128},
+                                   {0x52, 128},
+                                   {0x53, 128},
+                                   {0x60, 64},
+                                   {0x61, 128},
+                                   {-1, -1}};
+  int index = 0;
+  while (gpuArchCoresPerSM[index].SM != -1) {
+    if (gpuArchCoresPerSM[index].SM == ((major << 4) + minor))
+      return gpuArchCoresPerSM[index].Cores;
+    index++;
+  }
```

Comment:
This appears to be a purely host function.
If so, we can convert gpuArchCoresPerSM to static std::array (and remove the last element), and re-write this loop using `std::find`

(Again, not in this PR)

Reply:
This is a nice idea! I will do that and commit a PR.